backbone.js

```diff
@@ -1471,6 +1477,28 @@
       return this;
     },
 
+    // Determine if a registered handler matches the fragment when
+    // History emits a navigate event.
+    onNavigate: function(fragment) {
+      var matchedRoute = this.matchFragment(fragment);
+      if (!matchedRoute) return this;
+      var args = this._extractParameters(matchedRoute.route, fragment);
+      if (this.execute(matchedRoute.callback, args, matchedRoute.name) !== false) {
+        this.trigger.apply(this, ['route:' + matchedRoute.name].concat(args));
+        this.trigger('route', matchedRoute.name, args);
+      }
+      return this;
+    },
+
+    // Match a fragment with a registered handler
+    matchFragment: function(fragment) {
+      // If the root doesn't match, no routes can match either.
+      if (!Backbone.history.matchRoot()) return false;
```

Comment:
Oh, actually, you know...I think this might be worth discussing further. If history changes the URL, I can see one arguing that it should emit the navigation event **always** – even if you navigate away from the app's root. This seems simpler to document, and more predictable.

The alternative is defining the `navigate` event as behaving something like:

> The navigate event will be triggered whenever `history` navigates to a fragment, unless that fragment does not contain the application's URL root.

I suppose there could potentially be a *separate* event fired in that case...some sort of navigated-outside-the-app-root event. But this seems unnecessarily complex than it just always navigating.

After typing all of this up, it's making me wonder if the urlRoot even belongs in history, now. I should double-check where it's being used in history...